utils/src/lib.rs

```diff
@@ -204,3 +205,118 @@ pub fn encode_args(func: &Function, args: &[impl AsRef<str>]) -> Result<Vec<u8>>
     Ok(func.encode_input(&tokens)?)
 }
+
+/// Fetches a function signature given the selector using 4byte.directory
+pub async fn fourbyte(selector: &str) -> Result<Vec<(String, i32)>> {
+    #[derive(Deserialize)]
+    struct Decoded {
+        text_signature: String,
+        id: i32,
+    }
+
+    #[derive(Deserialize)]
+    struct ApiResponse {
+        results: Vec<Decoded>,
+    }
+
+    let selector = &selector.strip_prefix("0x").unwrap_or(selector);
+    if selector.len() < 8 {
+        return Err(eyre::eyre!("Invalid selector"))
+    }
+    let selector = &selector[..8];
+
+    let url = format!("https://www.4byte.directory/api/v1/signatures/?hex_signature={}", selector);
+    let res = reqwest::get(url).await?;
+    let api_response = res.json::<ApiResponse>().await?;
+
+    Ok(api_response
+        .results
+        .into_iter()
+        .map(|d| (d.text_signature, d.id))
+        .collect::<Vec<(String, i32)>>())
+}
+
+pub async fn fourbyte_possible_sigs(calldata: &str, id: Option<String>) -> Result<Vec<String>> {
+    let mut sigs = fourbyte(calldata).await?;
+
+    match id {
+        Some(id) => {
+            let sig = match &id[..] {
+                "earliest" => {
+                    sigs.sort_by(|a, b| a.1.cmp(&b.1));
+                    sigs.get(0)
+                }
+                "latest" => {
+                    sigs.sort_by(|a, b| b.1.cmp(&a.1));
+                    sigs.get(0)
+                }
+                _ => {
+                    let id: i32 = id.parse().expect("Must be integer");
+                    sigs = sigs
+                        .iter()
+                        .filter(|sig| sig.1 == id)
+                        .cloned()
+                        .collect::<Vec<(String, i32)>>();
+                    sigs.get(0)
+                }
+            };
+            match sig {
+                Some(sig) => Ok(vec![sig.clone().0]),
+                None => Ok(vec![]),
+            }
+        }
+        None => {
+            // filter for signatures that can be decoded
+            Ok(sigs
+                .iter()
+                .map(|sig| sig.clone().0)
+                .filter(|sig| {
+                    let res = abi_decode(sig, calldata, true);
+                    res.is_ok()
+                })
+                .collect::<Vec<String>>())
+        }
+    }
+}
+
+pub fn abi_decode(sig: &str, calldata: &str, input: bool) -> Result<Vec<Token>> {
+    let func = IntoFunction::into(sig);
+    let calldata = calldata.strip_prefix("0x").unwrap_or(calldata);
+    let calldata = hex::decode(calldata)?;
+    let res = if input {
+        // need to strip the function selector
+        func.decode_input(&calldata[4..])?
+    } else {
+        func.decode_output(&calldata)?
+    };
+
+    // in case the decoding worked but nothing was decoded
+    if res.is_empty() {
+        eyre::bail!("no data was decoded")
+    }
+
+    Ok(res)
+}
+
+#[tokio::test]
+async fn test_fourbyte() {
```

Comment:
These tests should be wrapped in a

```rust
#[cfg(test)]
mod tests {
       use super::*
       ...
}
```

block.

Reply:
done